src/sagemaker_xgboost_container/algorithm_mode/callback.py

```diff
@@ -0,0 +1,205 @@
+import logging
+import os
+import tempfile
+import threading
+import queue
+from typing import Callable, Optional
+from xgboost import rabit
+from xgboost.core import CallbackEnv
+from xgboost.callback import _fmt_metric as fmt_metric
+
+
+TEMP_FILE_SUFFIX = ".sagemaker-ignore"
+FILE_LOCK_SUFFIX = ".sagemaker-uploading"
+FILE_SAFE_SUFFIX = ".sagemaker-uploaded"
+
+CHECKPOINT_FILENAME = "xgboost-checkpoint"
+CHECKPOINT_NUM_DIGITS = 12
+
+logger = logging.getLogger(__name__)
+logging.basicConfig(level=logging.INFO)
+
+
+def _zero_pad(iteration: int) -> str:
+
+    iter_num = format(iteration, "0{}".format(CHECKPOINT_NUM_DIGITS))
+
+    return iter_num
+
+
+# modified from https://github.com/dmlc/xgboost/blob/master/python-package/xgboost/callback.py
+def print_evaluation(period=1, show_stdv=True, start_iteration=0):
+    """Create a callback that print evaluation result.
+
+    We print the evaluation results every **period** iterations
+    and on the first and the last iterations.
+
+    Parameters
+    ----------
+    period : int
+        The period to log the evaluation results
+
+    show_stdv : bool, optional
+         Whether show stdv if provided
+
+    Returns
+    -------
+    callback : function
+        A callback that print evaluation every period iterations.
+    """
+    def callback(env):
+        """internal function"""
+        if env.rank != 0 or (not env.evaluation_result_list) or period is False or period == 0:
+            return
+        i = env.iteration
+        if i % period == 0 or i + 1 == env.begin_iteration or i + 1 == env.end_iteration:
+            msg = '\t'.join([fmt_metric(x, show_stdv) for x in env.evaluation_result_list])
+            rabit.tracker_print('[%d]\t%s\n' % (i + start_iteration, msg))
+    return callback
+
+
+class SaveCheckpoint:
+    """Create a callback that saves checkpoints to disk.
+
+    This class represents the hook which is meant to be used a callback
+    function in XGBoost.
+
+    Example
+    -------
+    >>> save_checkpoint = SaveCheckpoint("/opt/ml/checkpoints")
+    >>> xgboost.train(prams, dtrain, callbacks=[save_checkpoint])
+
+    At the end of each iteration, we save each checkpoint to different files
+    parametrized by the iteration number, e.g., checkpoint.01, checkpoint.02,
+    etc. However, this could result in a large number of files to save locally
+    (and on S3 if used in SageMaker). To save disk space and reduce the amount
+    of data required to be downloaded on resumption, we retain only the N
+    (default 5, spcified by 'max_to_keep') most recent checkpoints.
```

Comment:
Based on my understanding, if we have checkpoints queued up, the upload would only happen on the most recent checkpoint. So we retain n - 1 most recent uploaded checkpoints the queued checkpoints that have not been uploaded yet.

If my understanding is incorrect please let me know.